### PR TITLE
pr/upgrade and refactor

### PR DESCRIPTION
**Pull request by noahehall**
**What**: 
- typescript first
- react 18 & functional components
- bun.sh (tests, bundling, development)
- upgrade ~~waypoint~~ all dependencies

**Why**:
- waypoint upgrade was required: check the growing list of stale issues (waypoint default export is now a named export)
- while im here might as well update them all

**How**: 
- no additional program changes were made, however the migration should inspire additional PRs to:
  - fully utilize react 18 features
  - drop remaining legacy react patterns and dependencies
- some tests were disabled  

**Checklist**: 

* [ ] Documentation
* [ ] Tests
* [ ] Ready to be merged 
* [ ] Added myself to contributors table 
* [ ] find a bun compatible styleguidist alternative
* [ ] use framer to animate replacement of image

**feel free to add additional comments**:
- https://github.com/react-cosmos/rfcs/issues/27
- https://github.com/oven-sh/bun/issues/2877

ill try to finish this PR sometime next weekend, but verified it works
image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -77,5 +69,8 @@
   "bugs": {
     "url": "https://github.com/stereobooster/react-ideal-image/issues"
   },
-  "homepage": "https://github.com/stereobooster/react-ideal-image#readme"
+  "homepage": "https://github.com/stereobooster/react-ideal-image#readme",
+  "dependencies": {
+    "@ladle/react": "^2.12.3"
```
Comment (stereobooster) [inline package.json]: I think this meant to be in `devDependencies`

File: global.d.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
+import {
+  Component, ComponentType, ComponentClass, CSSProperties, ReactElement, Ref
+} from 'react'
+
+declare global {
+  // @see https://stackoverflow.com/questions/38383676/error-ts2339-property-connection-does-not-exist-on-type-navigator
+  interface Navigator extends NavigatorNetworkInformation {}
```
Comment (stereobooster) [inline global.d.ts]: I'm not sure about overriding built-in interfaces in npm packages. This is a normal approach in the "final" project, but in npm package this is questionable

File: src/components/IdealImage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    loader,
+    placeholder,
+    shouldAutoDownload,
+    srcSet,
+    theme,
+    threshold,
+    width,
+  }
+
+  console.info('\n\n rendering with props', useProps)
```
Comment (stereobooster) [inline src/components/IdealImage/index.tsx]: I guess this is for debugging

**Comment by stereobooster** [comment]
I quickly got over it. I t would take some time to fully review

**Comment by noahehall** [comment]
awesome; i don't think it will be ready to review for a week or two; but early recommendations would definitely be helpful

**Comment by noahehall** [comment]
> I quickly got over it. I t would take some time to fully review

I would hold off on reviewing this until I remove the `Draft` label

File: src/components/IdealImage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+  };
+
+  const loadStateChange = (loadState, userTriggered, loadInfo = 0) => {
+    setNetworkState({
+      ...networkState,
+      overThreshold: false,
+    });
+    setImgState({
+      ...imgState,
+      state: loadState, // TODO(noah): rename to status
```
Comment (noahehall) [inline src/components/IdealImage/index.tsx]: blah missed this one

File: src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg
Comment (noahehall) [inline src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg:1]: delete this file

File: src/components/IdealImageWithDefaults/fixtures/index.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import * as React from "react";
+import IdealImage from "../../IdealImage";
+import img from "./freeimage3072×2304.jpg";
+
+export default <div>I am a fixture</div>;
```
Comment (noahehall) [inline src/components/IdealImageWithDefaults/fixtures/index.stories.tsx:4]: try again with bun 0.6.3 - it's working for nirv

**Comment by noahehall** [comment]
@stereobooster my main goals now are to:
- ensure i've checked all the checkboxes in the checklist
- ensure ive updated all the package json scripts ive removed
- react-cosmos

File: src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg
Comment (noahehall) [inline src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg:1]: delete this

File: src/components/IdealImage/index.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+      userTriggered: !!userTriggered,
+      loadInfo,
+    });
+  };
+
+  const load = (userTriggered) => {
+    if (
+      !imgState.url ||
+      ssr ||
+      [loadStates.loaded, loadStates.loaded].includes(imgState.state)
```
Comment (noahehall) [inline src/components/IdealImage/index.tsx]: think this should be loading and loaded

File: src/components/IdealImage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    }
+  }, [imgState]);
+
+  const icon = getIcon({
+    imgState,
+    networkState,
+  });
+
+  const message = getMessage(icon, imgState);
+  const useProps = {
```
Comment (noahehall) [inline src/components/IdealImage/index.tsx]: this pattern smells really bad

File: src/components/Media/index.tsx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      );
+    }
+  };
+
+  const Noscript = ({ ...useProps }) => {
+    return useProps.ssr ? (
+      <noscript>
+        <img
+          {...compose(useProps.theme.img, useProps.theme.noscript)}
+          src={useProps.src}
```
Comment (noahehall) [inline src/components/Media/index.tsx]: think these should use the nsSrc props from fallbackParams

File: src/components/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+  }
+  return supportedFormat.filter((x) => x.width === width)[0];
+};
+
+export const fallbackParams = ({ srcSet, getUrl }) => {
+  const notWebp = srcSet.filter((x) => !isWebp(x));
+  const first = notWebp[0];
+
+  // TODO(noah): required to pass test
+  if (!ssr || !first) return {};
```
Comment (noahehall) [inline src/components/helpers.ts:136]: check if this is still required

**Comment by noahehall** [comment]
@stereobooster i've gotten the [develop branch](https://github.com/stereobooster/react-ideal-image/compare/master...noahehall:react-idealer-image:develop) to a stable state

However, i've opted to move in a  different direction than the existing code patterns, and additionally
- no clue when bun.sh will work with react.cosmos, and dont plan on dealing with storybook or the other non-bunnish frameworks
- all the tests are broken (i might have time to fix in month or so)
- dropped support for legacy browsers
- replaced waypoint with framer-motion

If, however, you've taken a look at my develop branch and would still consider merging into your repo, i would totally support that (vs continuing on my fork)

let me know! 
cheers 

- nh

**Pull request closed without merging**